### ISSUE TITLE
Memory leak using numpyro backend

### ISSUE THREAD

**Issue by null-a** [pyro-ppl/brmp#67]
The following program:

```python
import pandas as pd
from brmp.numpyro_backend import backend as numpyro
from brmp import defm
df = pd.DataFrame({'y': [0.,1.,2.], 'x': pd.Series([1,2,3])})
generated_model = defm('y~x', df).generate(numpyro)
for _ in range(1000):
    generated_model.nuts(iter=1)
```

... consumes ever more memory when running. The same program using the pyro backend uses constant memory. (Against commit `5571d21`.)

**Comment by neerajprad** [comment]
@null-a : This just seems like a consequence of JAX caching compiled assets. Since it uses an LRU cache, at some point the memory consumption would flatten out. Let me know if this is a major issue, and we'll figure out if there is something we can do about it in NumPyro or JAX. This does however indicate that we are recompiling parts of the code afresh each time, which we can probably improve upon. I'll follow up on that separately.

**Comment by null-a** [comment]
@neerajprad: Thanks for looking into this! IIRC it came to my attention when I had a process (that ran inference many times) terminated with some kind of out of memory condition, which seems to go against the expectation that memory usage would flatten out at some point. Let me see if I can replicate that -- perhaps there was something else going on in addition to this.

It's no trouble for me to work around this issue, so it's not a high priority from my perspective.

**Comment by null-a** [comment]
> I had a process (that ran inference many times) terminated with some kind of out of memory condition

I inadvertently ran into this again. The shell of the terminated process showed `Killed` and a few other bits that look uninteresting. Running `dmesg -T | tail -n3` shows:

```
[Wed Nov 20 11:18:06 2019] Out of memory: Kill process 1679 (python) score 972 or sacrifice child
[Wed Nov 20 11:18:06 2019] Killed process 1679 (python) total-vm:25445416kB, anon-rss:15565088kB, file-rss:72176kB, shmem-rss:10244kB
[Wed Nov 20 11:18:06 2019] oom_reaper: reaped process 1679 (python), now anon-rss:0kB, file-rss:74732kB, shmem-rss:10244kB
```

I still don't know whether this is the cache (mentioned earlier) growing without bound, or something else.

**Comment by neerajprad** [comment]
@null-a - Is it for the same code as above? Could you share the link to your code, so that we can make sure that this doesn't point to some other issue that we have missed?

**Comment by null-a** [comment]
> Is it for the same code as above? 

No. (But I'll point to the code I *was* running below.) The original snippet (above) is what I hope is a minimal reproduction of the problem. To help confirm this, I've just run an updated version of that snippet twice, and both times it consumed all of the available memory and the machine became unusable. i.e. Memory usage didn't level out. (No termination by the OOM watch dog thingy this time though. Not sure why.) The updated snippet is included in the details section below.

The instance of the problem I [mentioned](https://github.com/pyro-ppl/brmp/issues/67#issuecomment-556042489) earlier arose when running the following at commit 5ee2eb6 with the diff below (in details) applied:

```bash
cd brmp/oed
python rds.py oed 2
```

However, having confirmed that the repro snippet will consume all available memory, I would guess that it does capture the underlying issue.

<details>

```python
import pandas as pd
from brmp.numpyro_backend import backend as numpyro
from brmp import define_model
from brmp.design import metadata_from_df
df = pd.DataFrame({'y': [0.,1.,2.], 'x': pd.Series([1,2,3])})
generated_model = define_model('y~x', metadata_from_df(df)).gen(numpyro)
data = generated_model.encode(df)
for i in range(1000000):
        generated_model.run_algo('nuts', data, iter=1, warmup=0, num_chains=1, seed=None)
```


```diff
diff --git a/brmp/oed/rds.py b/brmp/oed/rds.py
index 311052c..3cafd97 100644
--- a/brmp/oed/rds.py
+++ b/brmp/oed/rds.py
@@ -172,6 +172,7 @@ def main(name, M):
         response_col='y',
         participant_col='p',
         design_cols=['x', 'z'],
+        num_next_trial_runs=25,
         **kwargs)
 
     # Compute the Bayes factor. (We avoid defining the model
```
</details>

**Comment by neerajprad** [comment]
> However, having confirmed that the repro snippet will consume all available memory, I would guess that it does capture the underlying issue.

Thanks for snippet! Let me dig into this a bit and update.

**Comment by neerajprad** [comment]
@null-a - The good news is that there is no such issue with the master version of numpyro and jax, though I can very clearly see the memory usage going up per iteration in your example above. I think the likely culprit was due to a reference cycle in jax that got resolved in https://github.com/google/jax/pull/1721. We'll cut a new release for numpyro as soon as this fix is released upstream in jax.

**Comment by mattjj** [comment]
Just pushed jax 0.1.53 to pypi. (If we're ever slow at releases and you need one, just open an issue on our issue tracker. It only takes a few seconds, but we often forget to do it!)

**Comment by mattjj** [comment]
FYI after google/jax#1721 compilation cache entries will be deleted after the underlying Python callable is deleted, but the cache size can grow without bound. (Previously we had a limit on the number of cache entries, but we had to remove it in google/jax#1721 to support weakrefs.) We plan to write our own logic to limit the cache size (not just number of entries, by querying XLA for the size of each compiled program), likely with an LRU eviction policy, but we're not acting on that urgently because as far as we know no one is getting compilation cache OOMs.

If you know otherwise, i.e. if you ever see the compilation cache blowing up, let us know on the JAX issue tracker!

**Comment by neerajprad** [comment]
> likely with an LRU eviction policy, but we're not acting on that urgently because as far as we know no one is getting compilation cache OOMs.

@mattjj - Thanks for cutting a release! I found that this issue is completely fixed with https://github.com/google/jax/pull/1721, so the cache size shouldn't be a problem, at least in this case. I had earlier thought that JAX's cache holds the reference to the python callable (even if nothing else does), and that was causing the memory to steadily increase. This was probably the case earlier with `fastcache` but since JAX started using a  WeakDictionary (https://github.com/google/jax/pull/1605) and with the fix from #1721, it no longer keeps these items in the cache, and the current behavior is very intuitive. 🙂We will let you know if we face any issues with caching (I think you can continue to treat the issue of cache size lower priority)!

**Comment by neerajprad** [comment]
This should be fixed in numpyro, but we can follow up there if this continues to be an issue.
